src/ai/backend/manager/api/session.py

```diff
@@ -508,52 +509,70 @@ async def _create(request: web.Request, params: Any) -> web.Response:
-            except asyncio.TimeoutError:
-                resp['status'] = 'TIMEOUT'
-            else:
-                await asyncio.sleep(0.5)
-                async with root_ctx.db.begin_readonly() as conn:
+        async def kernelpullprogress(reporter: ProgressReporter) -> None:
+            progress = [0, 0]
+
+            async def _get_status(kernel_id):
+                async with root_ctx.db.begin() as conn:
                     query = (
                         sa.select([
+                            kernels.c.id,
                             kernels.c.status,
-                            kernels.c.service_ports,
                         ])
                         .select_from(kernels)
                         .where(kernels.c.id == kernel_id)
                     )
                     result = await conn.execute(query)
-                    row = result.first()
-                if row['status'] == KernelStatus.RUNNING:
-                    resp['status'] = 'RUNNING'
-                    for item in row['service_ports']:
-                        response_dict = {
-                            'name': item['name'],
-                            'protocol': item['protocol'],
-                            'ports': item['container_ports'],
-                        }
-                        if 'url_template' in item.keys():
-                            response_dict['url_template'] = item['url_template']
-                        if 'allowed_arguments' in item.keys():
-                            response_dict['allowed_arguments'] = item['allowed_arguments']
-                        if 'allowed_envs' in item.keys():
-                            response_dict['allowed_envs'] = item['allowed_envs']
-                        resp['servicePorts'].append(response_dict)
-                else:
-                    resp['status'] = row['status'].name
+
+                return result.first()
+
+            async def _update_progress(
+                app: web.Application,
+                source: AgentId,
+                event: KernelPullProgressEvent,
+            ) -> None:
+                # update both current and total
+                progress[0] = int(event.current_progress)
+                progress[1] = int(event.total_progress)
+
+            root_ctx.event_dispatcher.subscribe(KernelPullProgressEvent, request.app, _update_progress)
+            kernel_id = resp['sessionId']
+            while True:
+                result = await _get_status(kernel_id)
+                if result is None:
+                    continue
+                if result['status'] == KernelStatus.PREPARING:
+                    await reporter.update(0)
+                if result['status'] == KernelStatus.RUNNING:
+                    break
+                reporter.current_progress = progress[0]
+                reporter.total_progress = progress[1]
+                await reporter.update(0)
+            await asyncio.sleep(0.5)
```

Comment:
The above loop looks like a busy-wait loop.
Maybe is an indent missing here?

Reply:
Thank you, I missed it.